batchglm/train/numpy/base_glm/estimator.py

```diff
@@ -86,8 +94,95 @@ def train_sequence(self, training_strategy, **kwargs):
                         "overrding %s from training strategy with value %s with new value %s\n"
                         % (x, str(d[x]), str(kwargs[x]))
                     )
+            if "dispersion_smoothing" in d:
+                if d["dispersion_smoothing"] == "sctransform":
+                    if FDataGrid is None or NWSmoother is None:
+                        logger.error("Missing optional dependency scikit-fda.")
             self.train(**d, **kwargs)
             logger.debug("Training sequence #%d complete", idx + 1)
+            if "dispersion_smoothing" in d:
+                if d["dispersion_smoothing"] == "sctransform":
+                    self.perform_vst()
+                elif d["dispersion_smoothing"] == "edger":
+                    raise NotImplementedError()
+                elif d["dispersion_smoothing"] == "deseq2":
+                    raise NotImplementedError()
+
+    def perform_vst(
+        self,
+        theta_regularization: str = "od_factor",
+        use_geometric_mean: bool = True,
+        gmean_eps: float = 1.0,
+        bw_adjust: float = 1.5,
+    ):
+        logger.info("Performing Dispersion smoothing with flavour sctransform...")
+        # compute geometric log mean counts
+        if use_geometric_mean:
+            genes_log_gmean = np.log10(geometric_mean(self.model_container.x, axis=0, eps=gmean_eps))
+        else:
+            genes_log_gmean = np.log10(np.mean(self.model_container.x, axis=0))
+        if isinstance(genes_log_gmean, dask.array.core.Array):
+            genes_log_gmean = genes_log_gmean.compute()
+
+        # specify which kind of regularization is performed
+        scale_param = np.exp(self.model_container.theta_scale[0])  # TODO check if this is always correct
+        if theta_regularization == "log_theta":
+            dispersion_par = np.log10(scale_param)
+        elif theta_regularization == "od_factor":
+            dispersion_par = np.log10(1 + np.power(10, genes_log_gmean) / scale_param)
+        else:
+            raise ValueError(f"Unrecognized regularization method {theta_regularization}")
+        if isinstance(dispersion_par, dask.array.core.Array):
+            dispersion_par = dispersion_par.compute()
+
+        # downsample because KDE and bandwidth selection would take too long if performed on the entire dataset.
+        # It is sufficient to get a general idea of the data distribution
```

Comment:
Do we want to allow a no-sampling approach in the case of > 2000 features, just so that people can choose to do this?  I also understand if it's too complicated code-wise.

Reply:
2000 should be sufficient to get an idea of the distribution. More than that and the entire algorithm becomes incredibly slow. Remember that sctransform's VST is a non-parametric approach using KDE.